lib/spack/spack/binary_distribution.py

```diff
@@ -460,37 +467,63 @@ def relocate_package(spec, workdir, allow_root):
     """
     buildinfo = read_buildinfo_file(workdir)
     new_path = spack.store.layout.root
-    new_prefix = spack.paths.prefix
+    new_prefix = spec.prefix
+    new_rel_prefix = str(os.path.relpath(spec.prefix, new_path))
     old_path = buildinfo['buildpath']
-    old_prefix = buildinfo.get('spackprefix', '/not/in/buildinfo/dictionary')
+    old_spack_prefix = buildinfo.get('spackprefix', '/not/in/buildinfo/dictionary')
+    old_rel_prefix = buildinfo.get('relative_prefix', 'not/in/buildinfo/dictionary')
+    old_prefix = old_path+os.sep+old_rel_prefix
+    new_spack_prefix = spack.paths.prefix
     rel = buildinfo.get('relative_rpaths', False)
-    if rel:
-        return
 
-    tty.msg("Relocating package from",
+    tty.debug("Relocating package from install root ",
             "%s to %s." % (old_path, new_path))
+    tty.debug("Relocating package from install path ",
+            "%s to %s." % (old_prefix, new_prefix))
+    tty.debug("Relocating package from spack path",
+            "%s to %s." % (old_spack_prefix, new_spack_prefix))
+    path_names = set()
     path_names = set()
     for filename in buildinfo['relocate_textfiles']:
         path_name = os.path.join(workdir, filename)
         # Don't add backup files generated by filter_file during install step.
         if not path_name.endswith('~'):
             path_names.add(path_name)
-    relocate.relocate_text(path_names, oldpath=old_path,
-                           newpath=new_path, oldprefix=old_prefix,
-                           newprefix=new_prefix)
+    relocate.relocate_text(path_names,
+                           oldpath=old_path,
+                           newpath=new_path,
+                           oldprefix=old_prefix,
+                           newprefix=new_prefix,
+                           oldsprefix=old_spack_prefix,
+                           newsprefix=new_spack_prefix)
+
     # If the binary files in the package were not edited to use
     # relative RPATHs, then the RPATHs need to be relocated
+    new_prefix = str(spec.prefix)
+    path_names = set()
+    for filename in buildinfo['relocate_binaries']:
+        path_name = os.path.join(workdir, filename)
+        path_names.add(path_name)
+        if len(new_prefix) <= len(old_prefix):
+            relocate.replace_prefix_bin(path_name,
+                                        old_prefix,
+                                        new_prefix)
+            relocate.replace_prefix_bin(path_name,
+                                        old_path,
+                                        new_path)
+        else:
+            tty.warn('Cannot do a binary string replacement'
```

Comment:
This would seem to be serious enough for an exception? We can't complete the task, and don't know how to recover. Do we have a clean-up phase for failed relocation attempts to avoid leaving non-functional cruft around?

Reply:
In principal yes this should be an exception but you should be able to override it. For example a debug build will have the paths in the symbols.

Reply:
You mean it's OK not to be able to re-tool the debug symbols? Is there a way to tell? Do a first pass ignoring debug symbols where we have to succeed, maybe, then a second pass where failure is OK?